src/custom-element.js

```diff
@@ -117,96 +117,117 @@ export function stubElements(win) {
+  const widthAttr = element.getAttribute('width');
+  const heightAttr = element.getAttribute('height');
+  if (!widthAttr && !heightAttr) {
+    return Layout.CONTAINER;
   }
+  const sizesAttr = element.getAttribute('sizes');
+  if (!widthAttr || widthAttr == 'auto') {
+    return Layout.FIXED_HEIGHT;
+  }
+  return sizesAttr ? Layout.RESPONSIVE : Layout.FIXED;
+}
+
+
+/**
+ * Applies layout to the element. Visible for testing only.
+ * @param {!AmpElement} element
+ * @return {!Layout}
+ */
+export function applyLayout_(element) {
+  const layout = determineLayout(element);
+
   element.classList.add(getLayoutClass(layout));
   if (isLayoutSizeDefined(layout)) {
     element.classList.add('-amp-layout-size-defined');
   }
 
-  if (layout == Layout.FIXED || layout == Layout.FIXED_HEIGHT ||
-          layout == Layout.RESPONSIVE) {
-    let width = 0;
-    if (layout == Layout.FIXED_HEIGHT) {
-      if (widthAttr && widthAttr != 'auto') {
-        throw new Error('Expected width to be either absent or equal "auto" ' +
-            'for fixed-height layout: ' + widthAttr);
-      }
-    } else {
-      width = parseLength(widthAttr);
-      if (!width) {
-        throw new Error('Expected width to be available and be an ' +
-            'integer/length value: ' + widthAttr);
-      }
-    }
-    const height = parseLength(heightAttr);
-    if (!height) {
-      throw new Error('Expected height to be available and be an ' +
-          'integer/length value: ' + heightAttr);
-    }
-    if (layout == Layout.RESPONSIVE) {
-      if (getLengthUnits(width) != getLengthUnits(height)) {
-        throw new Error('Length units should be the same for width ' + width +
-            ' and height ' + height);
-      }
-      const sizer = element.ownerDocument.createElement('i-amp-sizer');
-      sizer.style.display = 'block';
-      sizer.style.paddingTop =
-          ((getLengthNumeral(height) / getLengthNumeral(width)) * 100) + '%';
-      element.insertBefore(sizer, element.firstChild);
-      element.sizerElement_ = sizer;
-    } else if (layout == Layout.FIXED_HEIGHT) {
-      element.style.height = height;
-    } else {
-      element.style.width = width;
-      element.style.height = height;
-    }
-  } else if (layout == Layout.FILL) {
+  // Handle layouts which don't care about width or height.
+  if (layout == Layout.FILL) {
     // Do nothing.
+    return layout;
   } else if (layout == Layout.CONTAINER) {
     // Do nothing. Elements themselves will check whether the supplied
     // layout value is acceptable. In particular container is only OK
     // sometimes.
+    return layout;
   } else if (layout == Layout.NODISPLAY) {
     element.style.display = 'none';
+    return layout;
+  }
+
+  // Now handle the remaining layouts, all of which end up computing
+  // or requiring width and height.
+  assert(layout == Layout.FIXED || layout == Layout.FIXED_HEIGHT ||
+	 layout == Layout.RESPONSIVE);
+
+  let widthAttr = element.getAttribute('width');
```

Comment:
@dvoytenko Thank you, I didn't know. I fixed this and also made it so it only reads the width attr if it really needs it (this should be an improvement over the original then, actually).